content/blog/2025-09-15-hash-fields-expiration/index.md

```diff
@@ -0,0 +1,225 @@
+
+* Some flags are short-lived experiments, meant to turn off automatically after a few seconds or minutes.
+
+* Others are long-term rollouts, where the toggle might remain valid for days or weeks.
+
+If you only had key-level expirations, you were forced to compromise.
+You could put all flags for a customer into a single hash, but then they would all share one expiration time.
+Or, you could store each flag in its own key so that each could expire independently — but this would explode the number of keys your system has to manage,
+hurting memory efficiency and increasing overhead for operations.
+
+The inability to set different TTLs per field meant developers either built complex cleanup processes outside of Valkey, or gave up flexibility in how expirations were handled.
+
+## Technical Challenges and Design Alternatives
+
+At first glance, adding field expirations might sound like a simple matter of storing timestamps per field.
+In order to understand the problem, it’s important to understand how expiration is normally handled in Valkey today.
+
+Valkey uses two complementary mechanisms to reclaim expired keys:
+
+* **Lazy expiration:** A key is deleted only when accessed. If you try to read or write an expired key, Valkey notices it has passed its TTL and deletes it immediately. This is cheap, but untouched keys can linger indefinitely and waste memory.
+
+* **Active expiration:** A background cron job runs 10 times per second, sampling a small set of keys with expiration. Expired keys are deleted until a time budget is reached, ensuring memory is reclaimed proactively without introducing latency spikes.
+
+For hash fields, we made a key design choice: we did not implement lazy expiration.
+Adding expiration checks to every [`HGET`](/commands/hget), [`HSET`](/commands/hset), or [`HDEL`](/commands/hdel) would have complicated the hot path of hash commands and risked performance regressions.
+Instead, we extended the active expiration job so that it can now also scan field-level expiration buckets, alongside top-level keys.
+
+This approach keeps expiration logic unified, predictable, and free from new latency cliffs.
+But it also introduced a challenge:
+how do we efficiently track and clean up expired fields inside hashes that may contain thousands (or millions) of entries,
+when only a small subset are volatile — and an even smaller fraction are expired at any given time?
+The challenge is balancing three conflicting goals:
+
+* **Keep memory usage small.** Hashes can grow to millions of fields, so metadata overhead must be minimal.
+
+* **Maintain fast lookups and updates.** Most hash commands run in `O(1)` time, and expiration tracking can’t change that.
+
+* **Reclaim memory efficiently.** The active expiration job is time-bounded, so we need to minimize wasted CPU cycles spent scanning unexpired fields.
+
+## Alternatives Considered
```

Comment:
Something to consider. Does the average reader care about options that were rejected? It's definitely an interesting deep dive, but there is so much more detail in the chosen implementation.

Reply:
I think it will help both answer trivial questions that will come to the reader's mind + provides some way of getting a better understanding of the problem. The sections are not long.